### DISCUSSION TITLE
Do you have different performance benchmarks on different GPU types compared to WebGL / native?

### DISCUSSION THREAD

**Discussion by grazder** [gpuweb/gpuweb#4918] [Q&A]
Hello! I am interested in using WebGPU to optimize my neural network on GPU. 
So I'm currently running my models on various user CPUs and my question is will I get a performance boost if I start using WebGPU.
Do you have different performance tests on different types of GPUs compared to WebGL/CPU?
I am also interested in the performance on integrated GPUs. Will there be gains compared to a basic CPU?

**Comment by grazder** [comment]
About integrated GPUs: for example "Intel UHD Graphics", "Intel(R) Iris(R) Xe Graphics", "Intel(R) HD Graphics", "Apple M..."

**Comment by gyagp** [comment]
There are some existing efforts on WebGPU to power AI usages, like TensorFlow.js, ONNX Runtime, TVM (especially WebLLM), Transformers.js (based on ONNX Runtime) and so on. Generally speaking, WebGPU perf is better than WebGL, as WebGPU has a dedicated compute pipeline, while WebGL has to leverage rendering pipeline to simulate the compute. If your models are not too small (e.g., smaller than 10MB), WebGPU (or GPU) is much faster than CPU (WASM) on the same client side devices, even for integrated GPUs. 
TensorFlow.js has [a benchmark suite](https://tfjs-benchmarks.web.app/local-benchmark/), and you can use that to compare the perf among WASM, WebGL and WebGPU. 
Here is [a collection of demos](https://intel.github.io/web-ai-showcase/) that showcase what WebGPU can do for WebAI nowadays. The demos include many great efforts in projects like Transformers.js, TVM, Chrome (Gemini Nano integration), ONNX Runtime and so on.

**Reply by vadimkantorov** [reply]
Curious, is there some compiler able to pre-generate fused WebGPU kernels for a given model? TVM? Or maybe does there exist a WebGPU backend for Triton lang?

**Reply by gyagp** [reply]
TVM is the only compiler that has WebGPU backend IIUC. It's a good idea to have a WebGPU backend for Triton.

**Comment by arjami25** [comment]
Yes, WebGPU offers a solid performance boost over WebGL and CPUs, even on integrated GPUs. Exact gains depend on the workload and GPU, but for neural nets, you’ll likely see better parallelism and efficiency.

**Comment by Frendysemon** [comment]
Yes, in most cases WebGPU should be faster than CPU, especially for neural networks. Even integrated GPUs can give a nice boost depending on the workload.
